File: proteus/patch/ProteusTuning.swift

extension Proteus {
  
  enum Tuning {

//    static let fileDataCount = 262 // 256 data bytes

    static let patchTruss = try! SinglePatchTruss("proteus.tuning", 256, params: parms.params(), initFile: "proteus1-tuning-init", createFileData: {
      sysexData($0, deviceId: 0)
    }, parseOffset: 5, pack: { bodyData, param, value in
      Proteus.pack(&bodyData, parm: param.p!, value: value)
    }, unpack: { bodyData, param in
      Proteus.unpack(bodyData, parm: param.p!)
    })
    
    static let patchTransform: MidiTransform = .single(throttle: 300, deviceId, .wholePatch({ editorVal, bodyData in
      [(.sysex(sysexData(bodyData, deviceId: UInt8(editorVal))), 100)]
    }))
    
    static func sysexData(_ bytes: [UInt8], deviceId: UInt8) -> [UInt8] {
      [0xf0, 0x18, 0x04, deviceId, 0x05] + bytes + [0xf7]
    }
    
    static let parms: [Parm] = 128.map {
      .p([.octave, .i($0 / 12), .note, .i($0 % 12)], p: $0, .max(8192))
    }


  }
}

File: proteus/proteus_tuning.js
//    const fileDataCount = 262 // 256 data bytes

const sysexData = (deviceId) => [0xf0, 0x18, 0x04, deviceId, 0x05, 'b', 0xf7]

const parms = (128).map(i => 
  [['octave', i / 12,'note', i % 12], { p: i, max: 8192 }]
)

const patchTruss = {
  single: "proteus.tuning", 
  bodyDataCount: 256, 
  parms: parms, 
  initFile: "proteus1-tuning-init", 
  createFileData: {
    sysexData($0, deviceId: 0)
  },
  parseOffset: 5, pack: { bodyData, param, value in
    Proteus.pack(&bodyData, parm: param.p!, value: value)
  },
  unpack: { bodyData, param in
    Proteus.unpack(bodyData, parm: param.p!)
  },
}

const patchTransform = {
  type: 'singleWholePatch',
  throttle: 300,
  patch: [[sysexData(deviceId), 100]],
}

module.exports = {
  patchTruss,
  patchTransform,
}